src/Spectre.libDataset/Dataset.h

```diff
@@ -0,0 +1,192 @@
+/*
+* Dataset.h
+* Data-owning implementation of Dataset interfaces.
+*
+Copyright 2017 Maciej Gamrat, Grzegorz Mrukwa
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#pragma once
+
+#include <vector>
+#include "IDataset.h"
+#include "Exception.h"
+
+namespace Spectre::libDataset
+{
+    /// <summary>
+    /// Data-owning implementation of dataset interfaces.
+    /// </summary>
+    template<typename DataType, typename SampleMetadata, typename DatasetMetadata>
+    class Dataset : public IDataset<DataType, SampleMetadata, DatasetMetadata>
+    {
+    public:
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Dataset"/> class.
+        /// </summary>
+        /// <param name="data">The data.</param>
+        /// <param name="sampleMetadata">The sample metadata.</param>
+        /// <param name="metadata">The metadata.</param>
+        Dataset(gsl::span<const DataType> data, gsl::span<const SampleMetadata> sampleMetadata, DatasetMetadata metadata) :
+            m_data(data.begin(), data.end()), m_sampleMetadata(sampleMetadata.begin(), sampleMetadata.end()), m_metadata(metadata)
+        {
+            if(data.size() == sampleMetadata.size())
```

Comment:
The comparison of assembly produced for each case has been presented here:
https://www.diffnow.com/?report=j1bf5
- The proposed solution contains only 2 jumps, while the current 3.
- Moreover, the proposed approach generates less instructions.

Nevertheless:
- Current code generates assembly that during execution of valid dataset goes through a one conditional and one unconditional jump which indeed goes into favor of better branch predictioning.

Therefore, it should indeed stay the way it is in my opinion.

*Checked on O2.*